File: implementations/python/strongly_connected_components.py
"""
Reference Implementation: Strongly Connected Components (SCC) & Condensation Graphs
Demonstrates:
1. Kosaraju's 2-Pass DFS Algorithm (Finishing Order + Transpose Graph).
2. Tarjan's 1-Pass DFS Algorithm (Discovery Times + Low-Link Values + Active Stack).
3. Condensation DAG Construction (Deduplicated Component Meta-Graph).
4. 2-Satisfiability (2-SAT) Solver via Implication Graph and SCC Decomposition.

Language: Python 3
"""

from typing import List, Optional
import unittest


def strongly_connected_components_kosaraju(graph: List[List[int]]) -> List[List[int]]:
    """Computes SCCs using Kosaraju's two-pass DFS algorithm.

    Time Complexity: O(V + E)
    Space Complexity: O(V + E)
    """
    n = len(graph)
    visited = [False] * n
    order = []

    def dfs1(u: int):
        visited[u] = True
        for v in graph[u]:
            if not visited[v]:
                dfs1(v)
        order.append(u)

    for u in range(n):
        if not visited[u]:
            dfs1(u)

    transpose = [[] for _ in range(n)]
    for u in range(n):
        for v in graph[u]:
            transpose[v].append(u)

    visited = [False] * n
    order.reverse()
    components = []

    def dfs2(u: int, comp: List[int]):
        visited[u] = True
        comp.append(u)
        for v in transpose[u]:
            if not visited[v]:
                dfs2(v, comp)

    for u in order:
        if not visited[u]:
            comp = []
            dfs2(u, comp)
            components.append(comp)

    return components


def strongly_connected_components_tarjan(graph: List[List[int]]) -> List[List[int]]:
    """Computes SCCs using Tarjan's single-pass DFS algorithm with discovery times and low-link values.

    Time Complexity: O(V + E)
    Space Complexity: O(V)
    """
    n = len(graph)
    disc = [-1] * n
    low = [-1] * n
    on_stack = [False] * n
    stack = []
    components = []
    timer = 0

    def dfs(u: int):
        nonlocal timer
        disc[u] = low[u] = timer
        timer += 1
        stack.append(u)
        on_stack[u] = True

        for v in graph[u]:
            if disc[v] == -1:
                # Tree edge
                dfs(v)
                low[u] = min(low[u], low[v])
            elif on_stack[v]:
                # Back edge to a node currently in the active DFS recursion stack
                low[u] = min(low[u], disc[v])

        # u is the root of an SCC
        if low[u] == disc[u]:
            comp = []
            while True:
                v = stack.pop()
                on_stack[v] = False
                comp.append(v)
                if v == u:
                    break
            components.append(comp)

    for u in range(n):
        if disc[u] == -1:
            dfs(u)

    return components


def make_component_id(n: int, components: List[List[int]]) -> List[int]:
    """Maps each vertex to its zero-indexed component identifier."""
    component_id = [-1] * n
    for cid, comp in enumerate(components):
        for u in comp:
            component_id[u] = cid
    return component_id


def build_condensation_dag(
    graph: List[List[int]],
    component_id: List[int],
    component_count: int,
) -> List[List[int]]:
    """Builds the condensation DAG of components with deduplicated cross-edges."""
    dag_set = [set() for _ in range(component_count)]

    for u in range(len(graph)):
        for v in graph[u]:
            cu = component_id[u]
            cv = component_id[v]
            if cu != cv:
                dag_set[cu].add(cv)

    return [sorted(neighbors) for neighbors in dag_set]


class TwoSatSolver:
    """Solves 2-Satisfiability problems via implication graphs and SCC decomposition."""

    def __init__(self, num_vars: int):
        self.num_vars = num_vars
        self.adj = [[] for _ in range(2 * num_vars)]

    def add_clause(self, u: int, val_u: bool, v: int, val_v: bool):
        """Adds clause: (u == val_u) OR (v == val_v)."""
        lit_u = 2 * u + (0 if val_u else 1)
        lit_v = 2 * v + (0 if val_v else 1)
        neg_u = lit_u ^ 1
        neg_v = lit_v ^ 1
        # ~lit_u => lit_v, ~lit_v => lit_u
        self.adj[neg_u].append(lit_v)
        self.adj[neg_v].append(lit_u)

    def solve(self) -> Optional[List[bool]]:
        """Finds a satisfying truth assignment if one exists, or returns None if unsatisfiable."""
        sccs = strongly_connected_components_tarjan(self.adj)
        comp_id = make_component_id(2 * self.num_vars, sccs)

        for i in range(self.num_vars):
            if comp_id[2 * i] == comp_id[2 * i + 1]:
                return None  # Contradiction: x and ~x belong to the same SCC

        # Tarjan produces components in reverse topological order.
        # comp_id[2*i] < comp_id[2*i+1] means literal x is finished earlier in reverse topological order
        # (topologically later / reachable from ~x), so assign True.
        assignment = [comp_id[2 * i] < comp_id[2 * i + 1] for i in range(self.num_vars)]
        return assignment


class TestStronglyConnectedComponents(unittest.TestCase):
    def _assert_same_partition(self, n: int, scc_a: List[List[int]], scc_b: List[List[int]]):
        id_a = make_component_id(n, scc_a)
        id_b = make_component_id(n, scc_b)
        for u in range(n):
            for v in range(n):
                self.assertEqual(
                    id_a[u] == id_a[v],
                    id_b[u] == id_b[v],
                    f"Partition mismatch between nodes {u} and {v}",
                )

    def test_two_scc_graph(self):
        # 0 -> 1 -> 2 -> 0, 2 -> 3, 3 -> 4 -> 5 -> 3
        g = [
            [1],
            [2],
            [0, 3],
            [4],
            [5],
            [3],
        ]
        scc_k = strongly_connected_components_kosaraju(g)
        scc_t = strongly_connected_components_tarjan(g)

        self.assertEqual(len(scc_k), 2)
        self.assertEqual(len(scc_t), 2)
        self._assert_same_partition(6, scc_k, scc_t)

        comp_id = make_component_id(6, scc_t)
        self.assertEqual(comp_id[0], comp_id[1])
        self.assertEqual(comp_id[1], comp_id[2])
        self.assertEqual(comp_id[3], comp_id[4])
        self.assertEqual(comp_id[4], comp_id[5])
        self.assertNotEqual(comp_id[0], comp_id[3])

        dag = build_condensation_dag(g, comp_id, 2)
        self.assertEqual(len(dag), 2)
        c_source = comp_id[0]
        c_sink = comp_id[3]
        self.assertEqual(dag[c_source], [c_sink])
        self.assertEqual(dag[c_sink], [])

    def test_linear_dag(self):
        # 0 -> 1 -> 2 -> 3
        g = [[1], [2], [3], []]
        scc_k = strongly_connected_components_kosaraju(g)
        scc_t = strongly_connected_components_tarjan(g)

        self.assertEqual(len(scc_k), 4)
        self.assertEqual(len(scc_t), 4)
        self._assert_same_partition(4, scc_k, scc_t)

    def test_single_cycle(self):
        # 0 -> 1 -> 2 -> 3 -> 4 -> 0
        g = [[1], [2], [3], [4], [0]]
        scc_k = strongly_connected_components_kosaraju(g)
        scc_t = strongly_connected_components_tarjan(g)

        self.assertEqual(len(scc_k), 1)
        self.assertEqual(len(scc_t), 1)
        self.assertEqual(len(scc_k[0]), 5)
        self.assertEqual(len(scc_t[0]), 5)
        self._assert_same_partition(5, scc_k, scc_t)

    def test_disconnected_graph(self):
        # 0 -> 1 -> 0, 2 (isolated), 3 -> 4 -> 5 -> 3
        g = [[1], [0], [], [4], [5], [3]]
        scc_k = strongly_connected_components_kosaraju(g)
        scc_t = strongly_connected_components_tarjan(g)

        self.assertEqual(len(scc_k), 3)
        self.assertEqual(len(scc_t), 3)
        self._assert_same_partition(6, scc_k, scc_t)

    def test_twosat_satisfiable(self):
        solver = TwoSatSolver(2)
        clauses = [
            (0, True, 1, True),
            (0, False, 1, True),
            (0, True, 1, False),
        ]
        for u, vu, v, vv in clauses:
            solver.add_clause(u, vu, v, vv)

        ans = solver.solve()
        self.assertIsNotNone(ans)
        self.assertEqual(len(ans), 2)
        for u, vu, v, vv in clauses:
            self.assertTrue((ans[u] == vu) or (ans[v] == vv))

    def test_twosat_unsatisfiable_single_var(self):
        solver = TwoSatSolver(1)
        solver.add_clause(0, True, 0, True)
        solver.add_clause(0, False, 0, False)
        ans = solver.solve()
        self.assertIsNone(ans)

    def test_twosat_xor_xnor_unsat(self):
        solver = TwoSatSolver(2)
        solver.add_clause(0, True, 1, True)
        solver.add_clause(0, False, 1, False)
        solver.add_clause(0, False, 1, True)
        solver.add_clause(0, True, 1, False)
        ans = solver.solve()
        self.assertIsNone(ans)

    def test_twosat_complex_sat(self):
        solver = TwoSatSolver(3)
        clauses = [
            (0, True, 1, True),
            (1, False, 2, True),
            (2, False, 0, False),
            (0, True, 2, False),
        ]
        for u, vu, v, vv in clauses:
            solver.add_clause(u, vu, v, vv)

        ans = solver.solve()
        self.assertIsNotNone(ans)
        for u, vu, v, vv in clauses:
            self.assertTrue((ans[u] == vu) or (ans[v] == vv))


if __name__ == "__main__":
    unittest.main()

File: implementations/cpp/strongly_connected_components.cpp
/**
 * Reference Implementation: Strongly Connected Components (SCC) & Condensation Graphs
 * Demonstrates:
 * 1. Kosaraju's 2-Pass DFS Algorithm (Finishing Order + Transpose Graph).
 * 2. Tarjan's 1-Pass DFS Algorithm (Discovery Times + Low-Link Values + Active Stack).
 * 3. Condensation DAG Construction (Deduplicated Component Meta-Graph).
 * 4. 2-Satisfiability (2-SAT) Solver via Implication Graph and SCC Decomposition.
 *
 * Language: C++17
 */

#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
#include <set>
#include <functional>
#include <cassert>

// ============================================================================
// 1. Kosaraju's Algorithm (2-Pass DFS via Transpose Graph)
// ============================================================================
std::vector<std::vector<int>> strongly_connected_components_kosaraju(
    const std::vector<std::vector<int>>& graph) {
    int n = static_cast<int>(graph.size());
    std::vector<int> visited(n, 0);
    std::vector<int> order;
    order.reserve(n);

    // Pass 1: Forward DFS to record finish order (postorder)
    std::function<void(int)> dfs1 = [&](int u) {
        visited[u] = 1;
        for (int v : graph[u]) {
            if (!visited[v]) {
                dfs1(v);
            }
        }
        order.push_back(u);
    };

    for (int u = 0; u < n; ++u) {
        if (!visited[u]) {
            dfs1(u);
        }
    }

    // Build transpose graph (reversed edges)
    std::vector<std::vector<int>> transpose(n);
    for (int u = 0; u < n; ++u) {
        for (int v : graph[u]) {
            transpose[v].push_back(u);
        }
    }

    // Pass 2: DFS on transpose in decreasing order of finish times
    std::fill(visited.begin(), visited.end(), 0);
    std::reverse(order.begin(), order.end());

    std::vector<std::vector<int>> components;

    std::function<void(int, std::vector<int>&)> dfs2 = [&](int u, std::vector<int>& comp) {
        visited[u] = 1;
        comp.push_back(u);
        for (int v : transpose[u]) {
            if (!visited[v]) {
                dfs2(v, comp);
            }
        }
    };

    for (int u : order) {
        if (!visited[u]) {
            std::vector<int> comp;
            dfs2(u, comp);
            components.push_back(std::move(comp));
        }
    }

    return components;
}

// ============================================================================
// 2. Tarjan's Algorithm (1-Pass DFS via Discovery Times, Low-Link, & Stack)
// ============================================================================
std::vector<std::vector<int>> strongly_connected_components_tarjan(
    const std::vector<std::vector<int>>& graph) {
    int n = static_cast<int>(graph.size());
    std::vector<int> disc(n, -1);
    std::vector<int> low(n, -1);
    std::vector<int> on_stack(n, 0);
    std::stack<int> st;
    std::vector<std::vector<int>> components;
    int timer = 0;

    std::function<void(int)> dfs = [&](int u) {
        disc[u] = low[u] = timer++;
        st.push(u);
        on_stack[u] = 1;

        for (int v : graph[u]) {
            if (disc[v] == -1) {
                // Tree edge: explore neighbor recursively
                dfs(v);
                low[u] = std::min(low[u], low[v]);
            } else if (on_stack[v]) {
                // Back edge to a vertex currently in the active DFS recursion stack
                low[u] = std::min(low[u], disc[v]);
            }
        }

        // u is the root of an SCC: pop vertices down to u
        if (low[u] == disc[u]) {
            std::vector<int> comp;
            while (true) {
                int v = st.top();
                st.pop();
                on_stack[v] = 0;
                comp.push_back(v);
                if (v == u) {
                    break;
                }
            }
            components.push_back(std::move(comp));
        }
    };

    for (int u = 0; u < n; ++u) {
        if (disc[u] == -1) {
            dfs(u);
        }
    }

    return components;
}

// ============================================================================
// 3. Component ID Mapping & Condensation DAG Construction
// ============================================================================
std::vector<int> make_component_id(
    int n,
    const std::vector<std::vector<int>>& components) {
    std::vector<int> component_id(n, -1);
    for (int cid = 0; cid < static_cast<int>(components.size()); ++cid) {
        for (int u : components[cid]) {
            component_id[u] = cid;
        }
    }
    return component_id;
}

std::vector<std::vector<int>> build_condensation_dag(
    const std::vector<std::vector<int>>& graph,
    const std::vector<int>& component_id,
    int component_count) {
    std::vector<std::set<int>> dag_set(component_count);

    for (int u = 0; u < static_cast<int>(graph.size()); ++u) {
        for (int v : graph[u]) {
            int cu = component_id[u];
            int cv = component_id[v];
            if (cu != cv) {
                dag_set[cu].insert(cv);
            }
        }
    }

    std::vector<std::vector<int>> dag(component_count);
    for (int c = 0; c < component_count; ++c) {
        dag[c] = std::vector<int>(dag_set[c].begin(), dag_set[c].end());
    }

    return dag;
}

// ============================================================================
// 4. 2-Satisfiability (2-SAT) Solver
// ============================================================================
class TwoSatSolver {
public:
    int num_vars;
    std::vector<std::vector<int>> adj;

    explicit TwoSatSolver(int vars) : num_vars(vars), adj(2 * vars) {}

    // Adds clause: (u == val_u) OR (v == val_v)
    void add_clause(int u, bool val_u, int v, bool val_v) {
        int lit_u = 2 * u + (val_u ? 0 : 1);
        int lit_v = 2 * v + (val_v ? 0 : 1);
        int neg_u = lit_u ^ 1;
        int neg_v = lit_v ^ 1;
        // ~lit_u => lit_v, ~lit_v => lit_u
        adj[neg_u].push_back(lit_v);
        adj[neg_v].push_back(lit_u);
    }

    // Solves 2-SAT instance. Returns true if satisfiable, and fills assignment.
    bool solve(std::vector<bool>& assignment) {
        assignment.assign(num_vars, false);
        std::vector<std::vector<int>> sccs = strongly_connected_components_tarjan(adj);
        std::vector<int> comp_id = make_component_id(2 * num_vars, sccs);

        for (int i = 0; i < num_vars; ++i) {
            if (comp_id[2 * i] == comp_id[2 * i + 1]) {
                return false; // Contradiction: x and ~x belong to the same SCC
            }
            // Tarjan produces components in reverse topological order.
            // If comp_id[2*i] < comp_id[2*i+1], literal x is finished earlier in reverse topological order,
            // which corresponds to x being topologically later (or reachable from ~x), so assign true.
            assignment[i] = (comp_id[2 * i] < comp_id[2 * i + 1]);
        }
        return true;
    }
};

// ============================================================================
// Unit Tests & Verification
// ============================================================================
static bool are_same_partition(int n,
                               const std::vector<std::vector<int>>& a,
                               const std::vector<std::vector<int>>& b) {
    std::vector<int> id_a = make_component_id(n, a);
    std::vector<int> id_b = make_component_id(n, b);

    // Two vertices u, v must be in the same component in a iff they are in the same component in b
    for (int u = 0; u < n; ++u) {
        for (int v = 0; v < n; ++v) {
            if ((id_a[u] == id_a[v]) != (id_b[u] == id_b[v])) {
                return false;
            }
        }
    }
    return true;
}

int main() {
    // ------------------------------------------------------------------------
    // Test 1: Standard 2-SCC Graph
    // 0 -> 1 -> 2 -> 0; 2 -> 3; 3 -> 4 -> 5 -> 3
    // ------------------------------------------------------------------------
    {
        std::vector<std::vector<int>> g(6);
        g[0] = {1};
        g[1] = {2};
        g[2] = {0, 3};
        g[3] = {4};
        g[4] = {5};
        g[5] = {3};

        auto scc_k = strongly_connected_components_kosaraju(g);
        auto scc_t = strongly_connected_components_tarjan(g);

        assert(scc_k.size() == 2);
        assert(scc_t.size() == 2);
        assert(are_same_partition(6, scc_k, scc_t));

        std::vector<int> comp_id = make_component_id(6, scc_t);
        assert(comp_id[0] == comp_id[1] && comp_id[1] == comp_id[2]);
        assert(comp_id[3] == comp_id[4] && comp_id[4] == comp_id[5]);
        assert(comp_id[0] != comp_id[3]);

        auto dag = build_condensation_dag(g, comp_id, 2);
        assert(dag.size() == 2);
        // The condensation DAG must have exactly one directed edge from {0,1,2} to {3,4,5}
        int c_source = comp_id[0];
        int c_sink = comp_id[3];
        assert(dag[c_source].size() == 1 && dag[c_source][0] == c_sink);
        assert(dag[c_sink].empty());
    }

    // ------------------------------------------------------------------------
    // Test 2: Linear DAG (Each vertex is its own SCC)
    // 0 -> 1 -> 2 -> 3
    // ------------------------------------------------------------------------
    {
        std::vector<std::vector<int>> g(4);
        g[0] = {1};
        g[1] = {2};
        g[2] = {3};

        auto scc_k = strongly_connected_components_kosaraju(g);
        auto scc_t = strongly_connected_components_tarjan(g);

        assert(scc_k.size() == 4);
        assert(scc_t.size() == 4);
        assert(are_same_partition(4, scc_k, scc_t));
    }

    // ------------------------------------------------------------------------
    // Test 3: Complete Directed Cycle (All vertices form 1 single SCC)
    // 0 -> 1 -> 2 -> 3 -> 4 -> 0
    // ------------------------------------------------------------------------
    {
        std::vector<std::vector<int>> g(5);
        g[0] = {1};
        g[1] = {2};
        g[2] = {3};
        g[3] = {4};
        g[4] = {0};

        auto scc_k = strongly_connected_components_kosaraju(g);
        auto scc_t = strongly_connected_components_tarjan(g);

        assert(scc_k.size() == 1);
        assert(scc_t.size() == 1);
        assert(scc_k[0].size() == 5);
        assert(scc_t[0].size() == 5);
        assert(are_same_partition(5, scc_k, scc_t));
    }

    // ------------------------------------------------------------------------
    // Test 4: Disconnected Graph with Multiple Isolated Components
    // 0 -> 1 -> 0, 2 (isolated), 3 -> 4, 4 -> 5, 5 -> 3
    // ------------------------------------------------------------------------
    {
        std::vector<std::vector<int>> g(6);
        g[0] = {1};
        g[1] = {0};
        g[3] = {4};
        g[4] = {5};
        g[5] = {3};

        auto scc_k = strongly_connected_components_kosaraju(g);
        auto scc_t = strongly_connected_components_tarjan(g);

        assert(scc_k.size() == 3);
        assert(scc_t.size() == 3);
        assert(are_same_partition(6, scc_k, scc_t));
    }

    // ------------------------------------------------------------------------
    // Test 5: 2-SAT Satisfiable Instance
    // (x0 or x1) and (~x0 or x1) and (x0 or ~x1)
    // ------------------------------------------------------------------------
    {
        TwoSatSolver solver(2);
        solver.add_clause(0, true, 1, true);   // (x0 or x1)
        solver.add_clause(0, false, 1, true);  // (~x0 or x1)
        solver.add_clause(0, true, 1, false);  // (x0 or ~x1)

        std::vector<bool> assignment;
        bool sat = solver.solve(assignment);
        assert(sat);
        assert(assignment.size() == 2);
        // Verify clauses
        assert(assignment[0] || assignment[1]);
        assert(!assignment[0] || assignment[1]);
        assert(assignment[0] || !assignment[1]);
        assert(assignment[0] == true && assignment[1] == true);
    }

    // ------------------------------------------------------------------------
    // Test 6: 2-SAT Unsatisfiable Contradiction
    // (x0 or x0) and (~x0 or ~x0) => x0 must be both true and false
    // ------------------------------------------------------------------------
    {
        TwoSatSolver solver(1);
        solver.add_clause(0, true, 0, true);
        solver.add_clause(0, false, 0, false);

        std::vector<bool> assignment;
        bool sat = solver.solve(assignment);
        assert(!sat);
    }

    // ------------------------------------------------------------------------
    // Test 7: 2-SAT XOR + XNOR Contradiction
    // (x0 ^ x1) AND (x0 == x1)
    // ------------------------------------------------------------------------
    {
        TwoSatSolver solver(2);
        // XOR: (x0 or x1) and (~x0 or ~x1)
        solver.add_clause(0, true, 1, true);
        solver.add_clause(0, false, 1, false);
        // XNOR: (~x0 or x1) and (x0 or ~x1)
        solver.add_clause(0, false, 1, true);
        solver.add_clause(0, true, 1, false);

        std::vector<bool> assignment;
        bool sat = solver.solve(assignment);
        assert(!sat);
    }

    // ------------------------------------------------------------------------
    // Test 8: 2-SAT 3-Variable Complex System
    // (x0 or x1), (~x1 or x2), (~x2 or ~x0), (x0 or ~x2)
    // ------------------------------------------------------------------------
    {
        TwoSatSolver solver(3);
        solver.add_clause(0, true, 1, true);
        solver.add_clause(1, false, 2, true);
        solver.add_clause(2, false, 0, false);
        solver.add_clause(0, true, 2, false);

        std::vector<bool> assignment;
        bool sat = solver.solve(assignment);
        assert(sat);
        // Verify all 4 clauses
        assert(assignment[0] || assignment[1]);
        assert(!assignment[1] || assignment[2]);
        assert(!assignment[2] || !assignment[0]);
        assert(assignment[0] || !assignment[2]);
    }

    std::cout << "[PASS] All Strongly Connected Components C++ unit tests passed.\n";
    return 0;
}
